opt: generalize LookupJoin operation

RaduBerinde opened the pull request:
The current LookupJoinOp is capable of expressing an index join, but
not a general join. Adjusting the private definition to identify the
key columns, and making the operator return all columns of the input
along with the looked up columns.

As a consequence, it is no longer possible to directly express an
index join where a PK column is not in the output; these cases now
require a renderNode (it is possible to collapse the renderNode in the
exec factory, but it's probably not worth it).

Release note: None

In a subsequent change I will split LookupJoin into LookupInnerJoin and LookupLeftJoin and add OnExpr.

RaduBerinde committed:
opt: generalize LookupJoin operation

The current LookupJoinOp is capable of expressing an index join, but
not a general join. Adjusting the private definition to identify the
key columns, and making the operator return all columns of the input
along with the looked up columns.

As a consequence, it is no longer possible to directly express an
index join where a PK column is not in the output; these cases now
require a renderNode (it is possible to collapse the renderNode in the
exec factory, but it's probably not worth it).

Release note: None

RaduBerinde committed:
opt: test updates

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26486)
<!-- Reviewable:end -->

justinj commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 19 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[pkg/sql/opt_exec_factory.go, line 352 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LELs8Fm6dvskjxxvsgN:-LELs8Fm6dvskjxxvsgO:bl6rt8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfe2f915777d19362c117f14627ff044a43fbaef/pkg/sql/opt_exec_factory.go#L352)):*
> ```Go
> }
> 
> // ConstructLookupJoin is part of the exec.Factory interface.
> ```

nit: comment still says LookupJoin

---

*[pkg/sql/opt/xform/explorer.go, line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LELscfQAbl7p0SlmUJT:-LELscfQAbl7p0SlmUJU:b-3i5tc) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfe2f915777d19362c117f14627ff044a43fbaef/pkg/sql/opt/xform/explorer.go#L261)):*
> ```Go
> 			input := e.f.ConstructScan(e.mem.InternScanOpDef(&indexScanOpDef))
> 
> 			// We scan whatever needed columns are left from the
> ```

unfinished comment?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-:-LELtYXbEwGXZGJT4mlg:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 19 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/opt_exec_factory.go, line 352 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LELs8Fm6dvskjxxvsgN:-LELw0dhBXOYpaCccE__:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfe2f915777d19362c117f14627ff044a43fbaef/pkg/sql/opt_exec_factory.go#L352)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: comment still says LookupJoin
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/explorer.go, line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LELscfQAbl7p0SlmUJT:-LELwYVE0WtQTpDPhIFc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfe2f915777d19362c117f14627ff044a43fbaef/pkg/sql/opt/xform/explorer.go#L261)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

unfinished comment?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26486)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/opt/xform/explorer.go, line 248 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEM-Ezy-RFyDxNoR9C2:-LELzjH66bL5hP4oFkyp:bb9izrf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dfa7a4b8453de5c5a581a080e2f2428449471d6/pkg/sql/opt/xform/explorer.go#L248)):*
> ```Go
> 			// We scan whatever columns we need which are available from the index,
> 			// plus the PK columns. The main reason is to allow pushing of filters as
> 			// much as possible.
> ```

Why is the main reason pushing of filters? Seems like the available cols + PK cols are required to make the index join logically equivalent to the original scan.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-:-LEM0Bnt0-9evBYZek-o:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/explorer.go, line 248 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEM-Ezy-RFyDxNoR9C2:-LEMeAuW9vEU6eYjVKkH:b-i008xc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dfa7a4b8453de5c5a581a080e2f2428449471d6/pkg/sql/opt/xform/explorer.go#L248)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why is the main reason pushing of filters? Seems like the available cols + PK cols are required to make the index join logically equivalent to the original scan.
</blockquote></details>

You only need the PK to do the lookup, and then you can retrieve whatever columns you want from the primary. So some columns, you could read from the secondary index or from the primary index. There isn't a huge benefit for one over the other except for the possibility of processing more filters before doing the lookup.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26486)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/metadata.go, line 129 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEMlbdaA80-owI1ubx1:-LEMlbdaA80-owI1ubx2:b5du605) ([raw file](https://github.com/cockroachdb/cockroach/blob/874f691e3fc1bd5a60e7059534d033333cce766d/pkg/sql/opt/metadata.go#L129)):*
> ```Go
> }
> 
> // IndexColumnsList returns the list of columns in the given index.
> ```

NIT: "...in the given index. The columns are returned in the same order as they appear in the index."

---

*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 542 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEMgtkJ5gBkZaOg8WV0:-LEMgtkJ5gBkZaOg8WV1:b-tl3qm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/874f691e3fc1bd5a60e7059534d033333cce766d/pkg/sql/opt/exec/execbuilder/relational_builder.go#L542)):*
> ```Go
> }
> 
> func (b *Builder) buildLookupJoin(ev memo.ExprView) (execPlan, error) {
> ```

Should this be called `buildLookupJoin` if it's now calling `ConstructIndexJoin`?

---

*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 545 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEMh36pAR3wONNbfO1Z:-LEMh36pAR3wONNbfO1_:b-icripn) ([raw file](https://github.com/cockroachdb/cockroach/blob/874f691e3fc1bd5a60e7059534d033333cce766d/pkg/sql/opt/exec/execbuilder/relational_builder.go#L545)):*
> ```Go
> 	// If lookup join child is a sort operator, then flip the order so that the
> 	// sort is on top of the lookup join.
> 	// TODO(radu): Remove this code once we have support for a more general
> ```

Is this comment still accurate? What impact does this PR have on it?

---

*[pkg/sql/opt/memo/expr_view.go, line 345 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEMhTe0DYG2PJ2Vy1gb:-LEMhTe0DYG2PJ2Vy1gc:b-8xrqat) ([raw file](https://github.com/cockroachdb/cockroach/blob/874f691e3fc1bd5a60e7059534d033333cce766d/pkg/sql/opt/memo/expr_view.go#L345)):*
> ```Go
> 		tableID := def.Table
> 		tp.Childf("table: %s", ev.Metadata().Table(tableID).TabName())
> 		tp.Childf("key cols: %v", def.KeyCols)
> ```

NIT: we use `columns` rather than `cols` for other props.

---

*[pkg/sql/opt/memo/private_defs.go, line 141 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEMiR1dCS0AGfEZShNo:-LEMiR1dCS0AGfEZShNp:bacg2i7) ([raw file](https://github.com/cockroachdb/cockroach/blob/874f691e3fc1bd5a60e7059534d033333cce766d/pkg/sql/opt/memo/private_defs.go#L141)):*
> ```Go
> 	Table opt.TableID
> 
> 	// KeyCols are the columns (produced by the input) used to create lookup keys;
> ```

I'd clarify that the column ids of these columns will be different than the column ids of the index. Maybe an example would be useful.

---

*[pkg/sql/opt/memo/private_defs.go, line 145 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEMied524jklbuMKa5-:-LEMied524jklbuMKa50:b4kqyxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/874f691e3fc1bd5a60e7059534d033333cce766d/pkg/sql/opt/memo/private_defs.go#L145)):*
> ```Go
> 	KeyCols opt.ColList
> 
> 	// Cols is the set of columns retrieved. The LookupJoin operator produces the
> ```

"... set of columns retrieved from the index."

---

*[pkg/sql/opt/memo/private_defs.go, line 147 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEMomus2iwDtYE9Nmw_:-LEMomus2iwDtYE9Nmwa:by3r33f) ([raw file](https://github.com/cockroachdb/cockroach/blob/874f691e3fc1bd5a60e7059534d033333cce766d/pkg/sql/opt/memo/private_defs.go#L147)):*
> ```Go
> 	// Cols is the set of columns retrieved. The LookupJoin operator produces the
> 	// columns in its input plus these columns.
> 	Cols opt.ColSet
> ```

Maybe this should be `ExtraCols`, or maybe `LookupCols`? There are so many columns floating around, i'm looking for naming that's a bit less generic. I could see:

`InputKey`: list of input columns used to create lookup keys; in same order as index columns; subset of input columns
`LookupCols`: set of columns retrieved from the index, this may include extra storage columns beyond the key columns

---

*[pkg/sql/opt/xform/explorer.go, line 230 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEMmURZDuGMA7n2tSW_:-LEMmURZDuGMA7n2tSWa:bicb4z4) ([raw file](https://github.com/cockroachdb/cockroach/blob/874f691e3fc1bd5a60e7059534d033333cce766d/pkg/sql/opt/xform/explorer.go#L230)):*
> ```Go
> 	tab := md.Table(scanOpDef.Table)
> 
> 	pkCols := md.IndexColumnsList(scanOpDef.Table, 0 /* indexOrdinal */)
> ```

NIT: better to use `opt.PrimaryIndex` rather than `0`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26486)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTRs

---

Review status: 2 of 20 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 542 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEMgtkJ5gBkZaOg8WV0:-LEN0SLVBekSi5SDARD9:bg6h4uo) ([raw file](https://github.com/cockroachdb/cockroach/blob/874f691e3fc1bd5a60e7059534d033333cce766d/pkg/sql/opt/exec/execbuilder/relational_builder.go#L542)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Should this be called `buildLookupJoin` if it's now calling `ConstructIndexJoin`?
</blockquote></details>

It will soon also call a new ConstructLookupJoin in the general case.

---

*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 545 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEMh36pAR3wONNbfO1Z:-LEN0aOhBWLS9aawGRVg:bji9da6) ([raw file](https://github.com/cockroachdb/cockroach/blob/874f691e3fc1bd5a60e7059534d033333cce766d/pkg/sql/opt/exec/execbuilder/relational_builder.go#L545)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Is this comment still accurate? What impact does this PR have on it?
</blockquote></details>

No impact, we don't support anything new on the execution side.

---

*[pkg/sql/opt/memo/private_defs.go, line 141 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEMiR1dCS0AGfEZShNo:-LEN1rQhCHHkOCK6JFJB:bm5ecxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/874f691e3fc1bd5a60e7059534d033333cce766d/pkg/sql/opt/memo/private_defs.go#L141)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'd clarify that the column ids of these columns will be different than the column ids of the index. Maybe an example would be useful.
</blockquote></details>

They can be the same when we're doing an index join. In the general case they will be different (they'd be coming from a different table). Added some examples.

---

*[pkg/sql/opt/memo/private_defs.go, line 145 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEMied524jklbuMKa5-:-LEN1UbO61HpijwFbQ8u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/874f691e3fc1bd5a60e7059534d033333cce766d/pkg/sql/opt/memo/private_defs.go#L145)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

"... set of columns retrieved from the index."
</blockquote></details>

Done.

---

*[pkg/sql/opt/memo/private_defs.go, line 147 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-LEMomus2iwDtYE9Nmw_:-LEN2YcV04C4dt1LHtDP:b-scxiwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/874f691e3fc1bd5a60e7059534d033333cce766d/pkg/sql/opt/memo/private_defs.go#L147)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Maybe this should be `ExtraCols`, or maybe `LookupCols`? There are so many columns floating around, i'm looking for naming that's a bit less generic. I could see:

`InputKey`: list of input columns used to create lookup keys; in same order as index columns; subset of input columns
`LookupCols`: set of columns retrieved from the index, this may include extra storage columns beyond the key columns

</blockquote></details>

Changed to LookupCols. I left KeyCols as is, InputKey sounds a bit ambiguous (it's not a key for the input, it's a key for something else formed from the input).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26486#-:-LEN6GEF1Lo9WrKP0eb-:b-814g6a)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
